### PR TITLE
GFORMS-2906 - Add the ability to look up a rate from a csv

### PR DESCRIPTION
**Pull request by cmanson**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/gform/csv/CsvTaxRateAdapter.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    val response: Option[TaxRate] = searchResult match {
+      case Some(list) if list.size == 1 => Some(list.head)
+      case _                            => None
+    }
+
+    response.map(rate => Json.toJson(rate))
+  }
+}
+
+case class TaxRate(regime: String, code: String, rate: String, startDate: LocalDate, endDate: LocalDate)
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/gform/csv/CsvTaxRateAdapter.scala:65]: `rate` must be a number

**Review by josef-vlach — CHANGES_REQUESTED** [review]

**Comment by cmanson** [comment]
Hi @josef-vlach, I've addressed all PR comments now I believe.

File: app/uk/gov/hmrc/gform/gform/csv/CsvTaxRateAdapter.scala [hunk 2]
```diff
@@ -33,36 +38,33 @@ class CsvTaxRateAdapter extends CsvDataRetrieveAdapter[TaxRate] {
...
-    val searchResult: Option[List[TaxRate]] = regimeRates.map(_.filter { rate =>
-      rate.code == taxRateRequest.code && (
-        rate.startDate.isEqual(requestedDate) || rate.startDate.isBefore(requestedDate)
-      ) && (
-        rate.endDate.isEqual(requestedDate) || rate.endDate.isAfter(requestedDate)
-      )
-    })
+  override def search(request: DataRetrieve.Request): Option[JsValue] =
+    Try(request.json.as[TaxRateRequest]) match {
+      case Success(taxRateRequest: TaxRateRequest) =>
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/gform/csv/CsvTaxRateAdapter.scala:49]: We should use `Try` only if there is no better alternative. In this case we can use `validate` [method](https://github.com/playframework/play-json/blob/main/play-json/shared/src/main/scala/play/api/libs/json/JsReadable.scala#L37)

```
request.json.validate[TaxRateRequest] match {
      case JsSuccess(taxRateRequest, _) =>
```

File: app/uk/gov/hmrc/gform/gform/csv/CsvTaxRateAdapter.scala [hunk 3]
```diff
@@ -33,36 +38,33 @@ class CsvTaxRateAdapter extends CsvDataRetrieveAdapter[TaxRate] {
...
+          rate.code === taxRateRequest.code && (
+            rate.startDate.isEqual(requestedDate) || rate.startDate.isBefore(requestedDate)
+          ) && (
+            rate.endDate.isEqual(requestedDate) || rate.endDate.isAfter(requestedDate)
+          )
+        })
+        response.map(rate => Json.toJson(rate))
+      case Failure(exception) =>
+        logger.error(s"An error occurred attempting to unmarshal request: ${request.json}")
+        throw exception
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/gform/csv/CsvTaxRateAdapter.scala:63]: Perhaps better alternative would be to do this:
```
case JsError(e) =>
  throw new Exception(
    s"An error occurred attempting to unmarshal request: ${request.json}. Error: ${Json.stringify(JsError.toJson(e))}"
  )
```

I think explicit logging is not needed here. We can rely on default error handler logging instead.

**Review by josef-vlach — CHANGES_REQUESTED** [review]

**Pull request merged**